### PR TITLE
New exchange: Mushino

### PR DESCRIPTION
**Pull request by andrewbomushino**
Mushino is a new derivatives exchange. On top of perpetuals and spot, they offer the world's first Bitcoin-settled index futures.

### CODE REVIEW

File: js/mushino.js [hunk 1]
```diff
@@ -0,0 +1,672 @@
...
+        const remaining = this.safeFloat (order, 'qty_remaining');
+        const filled = amount - remaining;
+        const cost = price * amount;
+        const id = this.safeString (order, 'id');
+        const type = this.safeString (order, 'type').toLowerCase ();
+        const side = this.safeString (order, 'side').toLowerCase ();
+        const fills = this.safeValue (order, 'fills');
+        const trades = this.parseTrades (fills || [], market);
+        let average = price;
+        if (trades && trades.length > 0) {
```
Comment (kroitor) [inline js/mushino.js]: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#working-with-array-lengths

File: js/mushino.js [hunk 2]
```diff
@@ -0,0 +1,672 @@
...
+        const response = await this.privateGetCompletedOrders (this.extend (request, params));
+        return this.parseOrders (response['result'], market, since, limit);
+    }
+
+    async fetchMyTrades (symbol = undefined, since = undefined, limit = undefined, params = {}) {
+        await this.loadMarkets ();
+        const request = {};
+        let market = undefined;
+        if (symbol !== undefined) {
+            market = this.market (symbol).id;
```
Comment (kroitor) [inline js/mushino.js]: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#accessing-dictionary-keys

File: js/mushino.js [hunk 3]
```diff
@@ -0,0 +1,672 @@
...
+        for (let i = 0; i < currencies.length; i++) {
+            const currencyId = currencies[i];
+            const balance = balances[currencyId];
+            const code = this.safeCurrencyCode (currencyId);
+            const account = this.account ();
+            const free_cash = this.safeFloat (balance, 'free');
+            const locked_cash = this.safeFloat (balance, 'locked');
+            const withdrawn_cash = this.safeFloat (balance, 'withdrawn');
+            const free_margin = this.safeFloat (balance, 'margin');
+            const free_total = free_cash + (free_margin || 0);
```
Comment (kroitor) [inline js/mushino.js]: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#adding-numbers-and-concatenating-strings

File: js/mushino.js [hunk 4]
```diff
@@ -0,0 +1,672 @@
...
+    }
+
+    parseTransaction (transaction, currency = undefined) {
+        const id = this.safeString (transaction, 'id');
+        const address = this.safeString (transaction, 'address');
+        const tx = this.safeValue (transaction, 'transaction');
+        const txHash = this.safeString (tx, 'hash');
+        const timestamp = this.safeInteger (transaction, 'time');
+        const currencyId = this.safeString (transaction, 'asset');
+        const type = id.indexOf (currencyId) !== -1 ? 'deposit' : 'withdrawal';
```
Comment (kroitor) [inline js/mushino.js]: We need to split this line into several lines and add some brackets around the ternary conditional.

**Comment by ttodua** [comment]
dead exchange

**Pull request closed without merging**